invenio_rdm_records/services/services.py

```diff
@@ -131,3 +148,229 @@ def oai_result_item(self, identity, oai_record_source):
+
+        # TODO try to find a duplicate request
+        req_cls = current_requests_service.record_cls
+        model_cls = req_cls.model_cls
+        requests = [
+            req_cls(rm.data, model=rm)
+            for rm in model_cls.query.filter(
+                model_cls.json["created_by"] == {"user": str(identity.id)},
+                model_cls.json["topic"] == {"record": id_},
+            )
+            if rm.data and rm.data["type"] == UserAccessRequest.type_id
+        ]
+
+        if any((req.is_open for req in requests)):
+            # TODO DuplicateRequestError
+            raise PermissionDeniedError()
+
+        record = self.record_cls.pid.resolve(id_)
+        data = {
+            "payload": {
+                "permission": record.parent.access.settings.default_access_permission,
+                "message": message,
+            }
+        }
+
+        # TODO might also be community managers
+        receiver = None
+        record_owner = record.parent.access.owner.resolve()
+        if record_owner:
+            receiver = record_owner
+
+        if receiver is None:
+            # TODO as fallback maybe admins?
+            pass
+
+        request = current_requests_service.create(
+            identity,
+            data,
+            UserAccessRequest,
+            receiver,
+            topic=record,
+            expires_at=None,  # TODO expiration date
+            expand=expand,
+            uow=uow,
+        )
+
+        # immediately submit the request, unless it has errors
+        # TODO is this correct?
+        if request.errors:
+            return request
+
+        message = {
+            "payload": {
+                "content": data["payload"].get("message") or "No comment",
+            }
+        }
+
+        # TODO can we pass the message here?
+        return current_requests_service.execute_action(
+            identity,
+            request.id,
+            "submit",
+            data=message,
+            uow=uow,
+        )
+
+    @unit_of_work()
+    def create_guest_access_request_token(
+        self, identity, id_, data, expand=False, uow=None
+    ):
+        """Create an request token that can be used to create an access request."""
+        # TODO detect duplicates
+        # TODO proper permission checks
+        if authenticated_user in identity.provides:
+            raise PermissionDeniedError("request_guest_access")
+
+        record = self.record_cls.pid.resolve(id_)
+        if current_app.config.get("MAIL_SUPPRESS_SEND", False):
+            # TODO log warning
+            print("email sending has been disabled!")
+
+        access_token = AccessRequestToken.create(
+            email=data["email"],
+            full_name=data["full_name"],
+            message=data["message"],
+            record_pid=id_,
+            shelf_life=timedelta(hours=6),
+        )
+
+        # TODO this doesn't work because it's creating an /api/ URL...
+        verify_url = url_for(
```

Comment:
I guess we need to find a solution for this, although not sure how 😅  I

Reply:
Potentially via parts of the machinery behind the **links templates**?
Haven't had time to look deeper into this yet though...